### PR TITLE
Migrate to hapi 17, node 8 and async/await closes #568

### LINKED ISSUE #568: hapi v17 support
What's the plan? Migrate? Deprecate?

### PR DESCRIPTION
**Pull request by corbinu**
Migrates to hapi 17, node 8 and async/await

I removed wreck support as wreck no longer issues the events needed

I am still trying to solve an issue where the tests emit ` MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 end listeners added. Use emitter.setMaxListeners() to increase limit`

Any comments and help would be appreciated

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -39,11 +38,8 @@ exports.register = (server, options, next) => {
...
-    return monitor.start(next);
+    monitor.start();
 };
 
 
-exports.register.attributes = {
-
-    pkg: require('../package.json')
-};
+exports.pkg = require('../package.json');
```
Comment (arb) [inline lib/index.js]: Why this change?

File: lib/monitor.js [hunk 2]
```diff
@@ -57,50 +57,22 @@ class Monitor {
...
-
     }
 
     startOps(interval) {
 
         this._ops && this._ops.start(interval);
     }
 
-    start(callback) {
+    start() {
```
[the remark this answers is not in the document]
Comment (corbinu) [reply lib/monitor.js]: Nice catch I also removed the use of arguments bellow that

File: lib/index.js (hunk 1 above)
Comment (corbinu) [reply lib/index.js]: The plugin format has changed without it you get this error
```
[1] "name" is required
      at Object.exports.apply (/root/node_modules/hapi/lib/config.js:22:10)
      at internals.Server.register (/root/node_modules/hapi/lib/server.js:384:31)
      at it (/root/test/index.js:70:22)
      at Immediate.setImmediate (/root/node_modules/lab/lib/runner.js:566:31)
      at runCallback (timers.js:800:20)
      at tryOnImmediate (timers.js:762:5)
      at processImmediate [as _immediateCallback] (timers.js:733:5)
```

Comment (frankthelen) [reply lib/index.js]: Hmmm, you're sure that this works? According to the error message you have to provide `exports.name` in addition.

Comment (corbinu) [reply lib/index.js]: @frankthelen Yes I am sure the error was provided simply by switching it back and and running the test suit to see it fail. pkg can be used as well please see the very last paragraph and code example here: https://hapijs.com/api

Comment (frankthelen) [reply lib/index.js]: Cool. You're right.

File: README.md [hunk 3]
```diff
@@ -54,19 +53,14 @@ const options = {
...
-    if (err) {
-        return console.error(err);
-    }
-    server.start(() => {
-        console.info(`Server started at ${ server.info.uri }`);
-    });
+await server.start();
 
-});
+console.info(`Server started at ${ server.info.uri }`);
```
Comment (arb) [inline README.md]: We should add something about the supported versions of hapi somewhere in the readme.

File: lib/monitor.js [hunk 4]
```diff
@@ -187,9 +154,10 @@ class Monitor {
             });
         }
 
-        return callback();
+        return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (arb) [reply lib/monitor.js]: I mean can you just not return anything? As you said, this isn't even an async function.

File: lib/monitor.js [hunk 5]
```diff
@@ -207,8 +169,9 @@ class Monitor {
 
         // Do a setImmediate here so that all the streams listening for "end" have a chance to run
         // https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L894-L897
-        setImmediate(callback);
+        return new Promise((resolve) => setImmediate(() => resolve()));
```
Comment (arb) [inline lib/monitor.js]: Curious if we still need this?

If so you can write it as `return new Promise((resolve) => setImmediate(resolve));`

Comment (corbinu) [reply lib/monitor.js]: I don't think so so removing

File: README.md (hunk 3 above)
Comment (corbinu) [reply README.md]: What format do you want this in. I looked at all the other plugins in the hapi namespace and didn't see an others doing this as was just going to follow whatever format is there.

Comment (corbinu) [reply README.md]: @arb Just wanted to check if you had a chance to think about this. Don't mean to rush know we all get busy just excited ;)

Comment (arb) [reply README.md]: Put it high in the README. Doesn't matter to me what the format is, just so long as it's there so when people open issues complaining, it is documented.

Comment (corbinu) [reply README.md]: @arb Done!

**Comment by teddmason** [comment]
I've just gone through upgrading a couple of apps to Hapi 17 and have also upgraded our good logging to @corbinu 's fork, all looking good for us, so far, thank you

**Comment by arb** [comment]
Thanks @corbinu for taking care of this migration!
